handlers/spendMsgHandler.go

```diff
@@ -0,0 +1,74 @@
+package handlers
+
+import (
+	"crypto/sha256"
+	"github.com/FourthState/plasma-mvp-sidechain/msgs"
+	"github.com/FourthState/plasma-mvp-sidechain/plasma"
+	"github.com/FourthState/plasma-mvp-sidechain/store"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"math/big"
+)
+
+// returns the next tx index in the current block
+type NextTxIndex func() uint16
+
+func NewSpendHandler(utxoStore store.UTXOStore, nextTxIndex NextTxIndex) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		spendMsg, ok := msg.(msgs.SpendMsg)
+		if !ok {
+			panic("Msg does not implement SpendMsg")
+		}
+
+		txIndex := nextTxIndex()
+		blockHeight := big.NewInt(ctx.BlockHeight())
+
+		// construct the confirmation hash
+		merkleHash := spendMsg.MerkleHash()
+		header := ctx.BlockHeader().DataHash
+		confirmationHash := sha256.Sum256(append(merkleHash, header...))
+
+		// positional part of these keys addeded when the new positions are created
+		var spenderKeys [][]byte
+		var positions []plasma.Position
+		positions = append(positions, plasma.NewPosition(blockHeight, txIndex, 0, nil))
+		spenderKeys = append(spenderKeys, append(spendMsg.Output0.Owner[:], positions[0].Bytes()...))
+		if spendMsg.HasSecondOutput() {
+			positions = append(positions, plasma.NewPosition(blockHeight, txIndex, 1, nil))
+			spenderKeys = append(spenderKeys, append(spendMsg.Output1.Owner[:], positions[1].Bytes()...))
+		}
+
+		var inputKeys [][]byte
+		for i, key := range spendMsg.GetSigners() {
```

Comment:
Redundant recover operation which is very expensive. We should have some way to cache results of GetSigner